File: new_practice_python/104.maximum-depth-of-binary-tree.py
#
# @lc app=leetcode id=104 lang=python3
#
# [104] Maximum Depth of Binary Tree
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    # def maxDepth(self, root: TreeNode) -> int:
    def maxDepth(self, root):
        if root is None:
            return 0

        level_cnt = 0
        level = [root]

        while level:
            level_cnt += 1

            next_level = []
            for ele in level:
                if ele.left is not None:
                    next_level.append(ele.left)
                if ele.right is not None:
                    next_level.append(ele.right)
                
            level = next_level

        return level_cnt


        

    

        

# @lc code=end

File: new_practice_cpp/104.maximum-depth-of-binary-tree.cpp
/*
 * @lc app=leetcode id=104 lang=cpp
 *
 * [104] Maximum Depth of Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxDepth(TreeNode* root) {
        int current_max = 0;
        helper(root, current_max, 0);
        return current_max;
    }

    void helper(TreeNode* root, int& current_max, int current_depth){
        if(!root){
            current_max = max(current_max, current_depth);
            // cout<<current_depth;
            return;
        }

        helper(root->left, current_max, current_depth+1);
        helper(root->right, current_max, current_depth+1);
    }
};
// @lc code=end

